Nomad job plan errors with the below logs.

<!--
Hi there,

Thank you for opening an issue. Please note that we try to keep the Nomad issue
tracker reserved for bug reports and feature requests. For general usage
questions, please see: https://www.nomadproject.io/community

-->

### Nomad version
1.3.4

### Operating system and Environment details
Debian ARM64

### Issue

When doing a nomad job plan, I am getting the following error
`Error submitting job: Put "http://100.66.210.90:4646/v1/jobs?namespace=default&region=home": EOF`
The plan was working if change_mode is NOT set to script.

### Reproduction steps
Added `change_mode="script"` and `change_script` to the job template stanza.

#### Expected Result
Plan works as expected.

#### Actual Result
Error submitting job: Put "http://100.66.210.90:4646/v1/jobs?namespace=default&region=home": EOF

### Job file (if appropriate)

Here is the snippet

```
          "Templates": [
            {
              "SourcePath": "",
              "DestPath": "local/Caddyfile",
              "EmbeddedTmpl": "<REDACTED>",
              "ChangeMode": "script",
              "ChangeSignal": "",
              "ChangeScript": null,
              "Splay": 5000000000,
              "Perms": "0644",
              "Uid": null,
              "Gid": null,
              "LeftDelim": "[[",
              "RightDelim": "]]",
              "Envvars": false,
              "VaultGrace": 0,
              "Wait": null
            }
          ],
```

### Nomad Server logs (if appropriate)

### Nomad Client logs (if appropriate)

```
Aug 28 19:24:36 rpi0 nomad[4032]:     2022-08-28T19:24:36.784+0530 [ERROR] http: http: panic serving 100.111.170.15:53910: runtime error: invalid memor>
Aug 28 19:24:36 rpi0 nomad[4032]: goroutine 2003 [running]:
Aug 28 19:24:36 rpi0 nomad[4032]: net/http.(*conn).serve.func1()
Aug 28 19:24:36 rpi0 nomad[4032]:         net/http/server.go:1850 +0xb8
Aug 28 19:24:36 rpi0 nomad[4032]: panic({0x2161220, 0x3e68400})
Aug 28 19:24:36 rpi0 nomad[4032]:         runtime/panic.go:890 +0x260
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad/structs.(*Template).Validate(0x4000794600)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/structs/structs.go:7825 +0x30c
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad/structs.(*Task).Validate(0x40017ee6e0, 0x400195a1b0, {0x400220e0d6, 0x6}, {0x400283c>
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/structs/structs.go:7297 +0x13c8
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad/structs.(*TaskGroup).Validate(0x4000bc07e0, 0x4000c42b40)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/structs/structs.go:6517 +0x21a0
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad/structs.(*Job).Validate(0x4000c42b40)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/structs/structs.go:4446 +0x1cdc
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad.jobValidate.Validate({}, 0x4000c42b40)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/job_endpoint_hooks.go:248 +0x38
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad.(*Job).admissionValidators(0x400083cc30, 0x46daf0?)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/job_endpoint_hooks.go:104 +0xf8
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad.(*Job).admissionControllers(0x40001161c0?, 0x2673a55?)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/job_endpoint_hooks.go:71 +0x44
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad.(*Job).Plan(0x400083cc30, 0x4001182a80, 0x4000783ce0)
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/job_endpoint.go:1606 +0xc4
Aug 28 19:24:36 rpi0 nomad[4032]: reflect.Value.call({0x4000862900?, 0x40001105c0?, 0x4e40?}, {0x266bf83, 0x4}, {0x40022493e8, 0x3, 0x4000783ce0?})
Aug 28 19:24:36 rpi0 nomad[4032]:         reflect/value.go:584 +0x688
Aug 28 19:24:36 rpi0 nomad[4032]: reflect.Value.Call({0x4000862900?, 0x40001105c0?, 0x251ea00?}, {0x40022493e8?, 0x251ea00?, 0x22a7c20?})
Aug 28 19:24:36 rpi0 nomad[4032]:         reflect/value.go:368 +0x90
Aug 28 19:24:36 rpi0 nomad[4032]: net/rpc.(*service).call(0x40001163c0, 0x2c3bd28?, 0x40022202c0?, 0x0, 0x40000b9e00, 0x257d540?, {0x251ea00?, 0x400118>
Aug 28 19:24:36 rpi0 nomad[4032]:         net/rpc/server.go:382 +0x20c
Aug 28 19:24:36 rpi0 nomad[4032]: net/rpc.(*Server).ServeRequest(0x4000f31578?, {0x2c3bd28, 0x40022202c0})
Aug 28 19:24:36 rpi0 nomad[4032]:         net/rpc/server.go:503 +0x110
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/nomad.(*Server).RPC(0x400099c380, {0x2673a55, 0x8}, {0x251ea00?, 0x4001182900}, {0x22a7c20>
Aug 28 19:24:36 rpi0 nomad[4032]:         github.com/hashicorp/nomad/nomad/server.go:1891 +0xf8
Aug 28 19:24:36 rpi0 nomad[4032]: github.com/hashicorp/nomad/command/agent.(*Agent).RPC(0x40011bf1d0?, {0x2673a55?, 0x4000cf6000?}, {0x251ea00?, 0x4001>
```

--- a/nomad/structs/structs.go
+++ b/nomad/structs/structs.go
@@ -7644,20 +7644,6 @@ var (
 	TemplateChangeModeInvalidError = errors.New("Invalid change mode. Must be one of the following: noop, signal, script, restart")
 )
 
-// ChangeScript holds the configuration for the script that is executed if
-// change mode is set to script
-type ChangeScript struct {
-	// Command is the full path to the script
-	Command string
-	// Args is a slice of arguments passed to the script
-	Args []string
-	// Timeout is the amount of seconds we wait for the script to finish
-	Timeout time.Duration
-	// FailOnError indicates whether a task should fail in case script execution
-	// fails or log script failure and don't interrupt the task
-	FailOnError bool
-}
-
 // Template represents a template configuration to be rendered for a given task
 type Template struct {
 	// SourcePath is the path to the template to be rendered
@@ -7739,6 +7725,10 @@ func (t *Template) Copy() *Template {
 		nt.Wait = t.Wait.Copy()
 	}
 
+	if t.ChangeScript != nil {
+		nt.ChangeScript = t.ChangeScript.Copy()
+	}
+
 	return nt
 }
 
@@ -7780,8 +7770,12 @@ func (t *Template) Validate() error {
 			_ = multierror.Append(&mErr, fmt.Errorf("cannot use signals with env var templates"))
 		}
 	case TemplateChangeModeScript:
-		if t.ChangeScript.Command == "" {
-			_ = multierror.Append(&mErr, fmt.Errorf("must specify script path value when change mode is script"))
+		if t.ChangeScript == nil {
+			_ = multierror.Append(&mErr, fmt.Errorf("must specify change script configuration value when change mode is script"))
+		}
+
+		if err = t.ChangeScript.Validate(); err != nil {
+			_ = multierror.Append(&mErr, err)
 		}
 	default:
 		_ = multierror.Append(&mErr, TemplateChangeModeInvalidError)
@@ -7822,6 +7816,44 @@ func (t *Template) DiffID() string {
 	return t.DestPath
 }
 
+// ChangeScript holds the configuration for the script that is executed if
+// change mode is set to script
+type ChangeScript struct {
+	// Command is the full path to the script
+	Command string
+	// Args is a slice of arguments passed to the script
+	Args []string
+	// Timeout is the amount of seconds we wait for the script to finish
+	Timeout time.Duration
+	// FailOnError indicates whether a task should fail in case script execution
+	// fails or log script failure and don't interrupt the task
+	FailOnError bool
+}
+
+func (cs *ChangeScript) Copy() *ChangeScript {
+	if cs == nil {
+		return nil
+	}
+
+	ncs := new(ChangeScript)
+	*ncs = *cs
+
+	return ncs
+}
+
+// Validate makes sure all the required fields of ChangeScript are present
+func (cs *ChangeScript) Validate() error {
+	if cs == nil {
+		return nil
+	}
+
+	if cs.Command == "" {
+		return fmt.Errorf("must specify script path value when change mode is script")
+	}
+
+	return nil
+}
+
 // WaitConfig is the Min/Max duration used by the Consul Template Watcher. Consul
 // Template relies on pointer based business logic. This struct uses pointers so
 // that we tell the different between zero values and unset values.
@@ -7839,11 +7871,11 @@ func (wc *WaitConfig) Copy() *WaitConfig {
 	nwc := new(WaitConfig)
 
 	if wc.Min != nil {
-		nwc.Min = &*wc.Min
+		nwc.Min = wc.Min
 	}
 
 	if wc.Max != nil {
-		nwc.Max = &*wc.Max
+		nwc.Max = wc.Max
 	}
 
 	return nwc

--- a/nomad/structs/structs.go
+++ b/nomad/structs/structs.go
@@ -45,6 +45,7 @@ import (
 	"github.com/miekg/dns"
 	"github.com/mitchellh/copystructure"
 	"golang.org/x/crypto/blake2b"
+	"golang.org/x/exp/slices"
 )
 
 var (
@@ -7721,13 +7722,8 @@ func (t *Template) Copy() *Template {
 	nt := new(Template)
 	*nt = *t
 
-	if t.Wait != nil {
-		nt.Wait = t.Wait.Copy()
-	}
-
-	if t.ChangeScript != nil {
-		nt.ChangeScript = t.ChangeScript.Copy()
-	}
+	nt.ChangeScript = t.ChangeScript.Copy()
+	nt.Wait = t.Wait.Copy()
 
 	return nt
 }
@@ -7837,6 +7833,9 @@ func (cs *ChangeScript) Copy() *ChangeScript {
 
 	ncs := new(ChangeScript)
 	*ncs = *cs
+
+	// args is a slice!
+	ncs.Args = slices.Clone(cs.Args)
 
 	return ncs
 }

--- a/nomad/structs/structs_test.go
+++ b/nomad/structs/structs_test.go
@@ -2586,16 +2586,19 @@ func TestTemplate_Copy(t *testing.T) {
 		SourcePath:   "/local/file.txt",
 		DestPath:     "/local/dest.txt",
 		EmbeddedTmpl: "tpl",
-		ChangeMode:   TemplateChangeModeSignal,
-		ChangeSignal: "SIGHUP",
-		Splay:        10 * time.Second,
-		Perms:        "777",
-		Uid:          pointer.Of(1000),
-		Gid:          pointer.Of(2000),
-		LeftDelim:    "[[",
-		RightDelim:   "]]",
-		Envvars:      true,
-		VaultGrace:   time.Minute,
+		ChangeMode:   TemplateChangeModeScript,
+		ChangeScript: &ChangeScript{
+			Command: "/bin/foo",
+			Args:    []string{"--force", "--debug"},
+		},
+		Splay:      10 * time.Second,
+		Perms:      "777",
+		Uid:        pointer.Of(1000),
+		Gid:        pointer.Of(2000),
+		LeftDelim:  "[[",
+		RightDelim: "]]",
+		Envvars:    true,
+		VaultGrace: time.Minute,
 		Wait: &WaitConfig{
 			Min: pointer.Of(time.Second),
 			Max: pointer.Of(time.Minute),
@@ -2606,8 +2609,9 @@ func TestTemplate_Copy(t *testing.T) {
 	t1.SourcePath = "/local/file2.txt"
 	t1.DestPath = "/local/dest2.txt"
 	t1.EmbeddedTmpl = "tpl2"
-	t1.ChangeMode = TemplateChangeModeRestart
-	t1.ChangeSignal = ""
+	t1.ChangeMode = TemplateChangeModeSignal
+	t1.ChangeScript.Command = "/bin/foobar"
+	t1.ChangeScript.Args = []string{"--forces", "--debugs"}
 	t1.Splay = 5 * time.Second
 	t1.Perms = "700"
 	t1.Uid = pointer.Of(5000)
@@ -2623,7 +2627,8 @@ func TestTemplate_Copy(t *testing.T) {
 	require.NotEqual(t, t1.DestPath, t2.DestPath)
 	require.NotEqual(t, t1.EmbeddedTmpl, t2.EmbeddedTmpl)
 	require.NotEqual(t, t1.ChangeMode, t2.ChangeMode)
-	require.NotEqual(t, t1.ChangeSignal, t2.ChangeSignal)
+	require.NotEqual(t, t1.ChangeScript.Command, t2.ChangeScript.Command)
+	require.NotEqual(t, t1.ChangeScript.Args, t2.ChangeScript.Args)
 	require.NotEqual(t, t1.Splay, t2.Splay)
 	require.NotEqual(t, t1.Perms, t2.Perms)
 	require.NotEqual(t, t1.Uid, t2.Uid)
@@ -2757,6 +2762,33 @@ func TestTemplate_Validate(t *testing.T) {
 					Min: pointer.Of(5 * time.Second),
 					Max: pointer.Of(10 * time.Second),
 				},
+			},
+			Fail: false,
+		},
+		{
+			Tmpl: &Template{
+				SourcePath:   "foo",
+				DestPath:     "local/foo",
+				ChangeMode:   "script",
+				ChangeScript: nil,
+			},
+			Fail: true,
+		},
+		{
+			Tmpl: &Template{
+				SourcePath:   "foo",
+				DestPath:     "local/foo",
+				ChangeMode:   "script",
+				ChangeScript: &ChangeScript{Command: ""},
+			},
+			Fail: true,
+		},
+		{
+			Tmpl: &Template{
+				SourcePath:   "foo",
+				DestPath:     "local/foo",
+				ChangeMode:   "script",
+				ChangeScript: &ChangeScript{Command: "/bin/foo"},
 			},
 			Fail: false,
 		},

